packages/shared/src/components/PostOptionsReadingHistoryMenu.tsx

```diff
@@ -0,0 +1,168 @@
+import ShareIcon from '../../icons/share.svg';
+import BookmarkIcon from '../../icons/bookmark.svg';
+import BookmarkIconFilled from '../../icons/filled/bookmark_filled.svg';
+import XIcon from '../../icons/x.svg';
+import useBookmarkPost from '../hooks/useBookmarkPost';
+import AuthContext from '../contexts/AuthContext';
+import { ReadHistoryInfiniteData } from '../hooks/useInfiniteReadingHistory';
+import { useCopyLink } from '../hooks/useCopyLink';
+import { MenuIcon } from './MenuIcon';
+import { QueryIndexes } from '../hooks/useReadingHistory';
+import { useShareOrCopyLink } from '../hooks/useShareOrCopyLink';
+import AnalyticsContext from '../contexts/AnalyticsContext';
+import { postAnalyticsEvent } from '../lib/feed';
+
+const PortalMenu = dynamic(() => import('./fields/PortalMenu'), {
+  ssr: false,
+});
+
+export type PostOptionsReadingHistoryMenuProps = {
+  post: ReadHistoryPost;
+  onHiddenMenu?: () => unknown;
+  displayCopiedMessageFunc?: () => void;
+  onHideHistoryPost?: (postId: string) => unknown;
+  indexes: QueryIndexes;
+};
+
+/* eslint-disable no-param-reassign */
+const updateReadingHistoryPost =
+  (
+    queryClient: QueryClient,
+    historyQueryKey: QueryKey,
+    readHistoryPostUpdated: Partial<ReadHistoryPost>,
+    { page, edge }: QueryIndexes,
+  ): ((args: { id: string }) => Promise<() => void>) =>
+  async () => {
+    const oldReadingHistory =
+      queryClient.getQueryData<ReadHistoryInfiniteData>(historyQueryKey);
+    const newItems = [...oldReadingHistory.pages];
+    const history = newItems[page].readHistory.edges[edge].node.post;
+    newItems[page].readHistory.edges[edge].node.post = {
+      ...history,
+      ...readHistoryPostUpdated,
+    };
+    queryClient.setQueryData<ReadHistoryInfiniteData>(historyQueryKey, {
+      ...oldReadingHistory,
+      pages: newItems,
+    });
+    return () => {
+      queryClient.setQueryData<ReadHistoryInfiniteData>(
+        historyQueryKey,
+        oldReadingHistory,
+      );
+    };
+  };
+
+const getBookmarkIconAndMenuText = (bookmarked: boolean) => {
+  if (bookmarked) {
+    return (
+      <>
+        <MenuIcon Icon={BookmarkIconFilled} />
+        Remove from bookmarks
+      </>
+    );
+  }
+  return (
+    <>
+      <MenuIcon Icon={BookmarkIcon} />
+      Save to bookmarks
+    </>
+  );
+};
+
+export default function PostOptionsReadingHistoryMenu({
+  post,
+  onHiddenMenu,
+  onHideHistoryPost,
+  indexes,
+}: PostOptionsReadingHistoryMenuProps): ReactElement {
+  const { user } = useContext(AuthContext);
+  const { trackEvent } = useContext(AnalyticsContext);
+  const queryClient = useQueryClient();
+  const historyQueryKey = ['readHistory', user?.id];
+
+  const [, copyLink] = useCopyLink(() => post.commentsPermalink);
+
+  const trackEventFunc = (eventName: string) => {
+    return () =>
+      trackEvent(
+        postAnalyticsEvent(eventName, post, {
+          extra: { origin: 'reading history post menu' },
```

Comment:
Aligned with the existing naming

```suggestion
          extra: { origin: 'reading history context menu' },
```